src/binary_trees.rs

```diff
@@ -13,26 +13,42 @@ use typed_arena::Arena;
 use rayon::prelude::*;
 
 struct Tree<'a> {
-    l: Option<&'a Tree<'a>>,
-    r: Option<&'a Tree<'a>>,
-    i: i32,
+    children: Option<(&'a Tree<'a>, &'a Tree<'a>)>,
+    item: i32,
 }
 
-fn item_check(t: Option<&Tree>) -> i32 {
-    match t {
-        None => 0,
-        Some(&Tree { l: None, r: None, i }) => i,
-        Some(&Tree { l, r, i }) => i + item_check(l) - item_check(r)
+fn item_check(tree: &Tree) -> i32 {
+    // This mutually tail recursive definition is significantly more efficient than the non-tail
+    // recursive one.
+    fn plus(sum: i32, tree: &Tree) -> i32 {
+        if let Some((left, right)) = tree.children {
+            minus(plus(sum + tree.item, left), right)
+        } else {
+            sum + tree.item
+        }
+    }
+    fn minus(sum: i32, tree: &Tree) -> i32 {
+        if let Some((left, right)) = tree.children {
+            plus(minus(sum - tree.item, left), right)
+        } else {
+            sum - tree.item
+        }
+    }
+    if let Some((left, right)) = tree.children {
+        minus(plus(tree.item, left), right)
+    } else {
+        tree.item
     }
 }
 
-fn bottom_up_tree<'r>(arena: &'r Arena<Tree<'r>>, i: i32, depth: i32) -> Option<&'r Tree<'r>> {
-    let (l, r) = if depth > 0 {
-        (bottom_up_tree(arena, 2 * i - 1, depth - 1), bottom_up_tree(arena, 2 * i, depth - 1))
-    } else {
-        (None, None)
-    };
-    Some(arena.alloc(Tree { l: l, r: r, i: i }))
+fn bottom_up_tree<'r>(arena: &'r Arena<Tree<'r>>, item: i32, depth: i32)
+                  -> &'r Tree<'r> {
+    let mut tree = arena.alloc(Tree { children: None, item: item });
```

Comment:
By allocating the node first, the arena is populated closer to iteration order: if we allocate the node at the end then the arena is first populated with all the leaves, and then with parents.